File: Java/Ape4Ejercicio.java
import java.util.Scanner;

// ==========================================
// CLASE: Estudiante
// ==========================================
class Estudiante {
    // Atributos encapsulados del estudiante
    private String cedula;
    private String nombre;
    private String apellido;
    private double nota1;
    private double nota2;
    private double nota3;
    private double promedio;
    private String estado;

    // Constructor por defecto necesario para inicializar el arreglo en memoria
    public Estudiante() {
        this.cedula = "";
        this.nombre = "";
        this.apellido = "";
        this.nota1 = 0.0;
        this.nota2 = 0.0;
        this.nota3 = 0.0;
        this.promedio = 0.0;
        this.estado = "Reprobado";
    }

    // Constructor parametrizado para asignar valores directamente
    public Estudiante(String cedula, String nombre, String apellido, double nota1, double nota2, double nota3) {
        this.cedula = cedula;
        this.nombre = nombre;
        this.apellido = apellido;
        this.nota1 = nota1;
        this.nota2 = nota2;
        this.nota3 = nota3;
        calcularPromedio();
        determinarEstado();
    }

    // Métodos Getters y Setters para acceso seguro a los atributos
    public String getCedula() { return cedula; }
    public void setCedula(String cedula) { this.cedula = cedula; }

    public String getNombre() { return nombre; }
    public void setNombre(String nombre) { this.nombre = nombre; }

    public String getApellido() { return apellido; }
    public void setApellido(String apellido) { this.apellido = apellido; }

    public double getNota1() { return nota1; }
    public void setNota1(double nota1) { this.nota1 = nota1; calcularPromedio(); determinarEstado(); }

    public double getNota2() { return nota2; }
    public void setNota2(double nota2) { this.nota2 = nota2; calcularPromedio(); determinarEstado(); }

    public double getNota3() { return nota3; }
    public void setNota3(double nota3) { this.nota3 = nota3; calcularPromedio(); determinarEstado(); }

    public double getPromedio() { return promedio; }
    public String getEstado() { return estado; }

    // REQUERIMIENTO 3: Calcular automáticamente el promedio
    public void calcularPromedio() {
        this.promedio = (nota1 + nota2 + nota3) / 3.0;
    }

    // ESPECIFICACIÓN: El estudiante aprueba si el promedio es >= 7.00
    public void determinarEstado() {
        if (this.promedio >= 7.00) {
            this.estado = "Aprobado";
        } else {
            this.estado = "Reprobado";
        }
    }

    // Imprime de forma ordenada los datos individuales del estudiante
    public void mostrarInformacion() {
        // Se usa String.format para limitar los decimales del promedio a dos en la salida de consola
        System.out.println("Cedula: " + cedula + " | Alumno: " + apellido + " " + nombre 
             + " | Notas: [" + nota1 + ", " + nota2 + ", " + nota3 + "]"
             + " | Promedio: " + String.format("%.2f", promedio) + " | Estado: " + estado);
    }
}

// ==========================================
// CLASE PRINCIPAL: Main
// ==========================================
public class Ape4Ejercicio {

    // REQUERIMIENTO 7: Validar que las notas estén estrictamente entre 0 y 10
    public static double solicitarNotaValidada(Scanner teclado, String numeroNota) {
        double nota;
        while (true) {
            System.out.print("  Ingrese la " + numeroNota + " (0 - 10): ");
            if (teclado.hasNextDouble()) {
                nota = teclado.nextDouble();
                if (nota >= 0.0 && nota <= 10.0) {
                    return nota; // Retorna la nota solo si es válida
                } else {
                    System.out.println("  [Error] Nota invalida. Debe ser un numero entre 0 y 10. Intente de nuevo.");
                }
            } else {
                System.out.println("  [Error] Entrada no valida. Debe ingresar un numero. Intente de nuevo.");
                teclado.next(); // Limpia la entrada incorrecta (letras, símbolos) para evitar bucles infinitos
            }
        }
    }

    public static void main(String[] args) {
        Scanner teclado = new Scanner(System.in);
        
        // REQUERIMIENTO 1: Registrar mínimo 5 estudiantes
        final int TOTAL_ESTUDIANTES = 5;
        Estudiante[] listaEstudiantes = new Estudiante[TOTAL_ESTUDIANTES];

        int aprobados = 0;
        int reprobados = 0;

        System.out.println("=== SISTEMA DE REGISTRO DE ESTUDIANTES (JAVA) ===\n");

        // Bucle para el ingreso de datos de cada uno de los estudiantes
        for (int i = 0; i < TOTAL_ESTUDIANTES; i++) {
            System.out.println("Estudiante #" + (i + 1) + ":");
            System.out.print("  Cedula: ");
            String cedula = teclado.next();
            System.out.print("  Nombre: ");
            String nombre = teclado.next();
            System.out.print("  Apellido: ");
            String apellido = teclado.next();

            // REQUERIMIENTO 2: Ingresar las 3 notas usando el método de validación
            double n1 = solicitarNotaValidada(teclado, "nota 1");
            double n2 = solicitarNotaValidada(teclado, "nota 2");
            double n3 = solicitarNotaValidada(teclado, "nota 3");
            System.out.println("-------------------------------------------");

            // Instanciación del objeto y almacenamiento en el arreglo de Java
            listaEstudiantes[i] = new Estudiante(cedula, nombre, apellido, n1, n2, n3);
        }

        // REQUERIMIENTO 4: Mostrar el listado completo de estudiantes registrados
        System.out.println("\n=== LISTADO COMPLETO DE ESTUDIANTES ===");
        for (int i = 0; i < TOTAL_ESTUDIANTES; i++) {
            listaEstudiantes[i].mostrarInformacion();

            // Conteo de estudiantes según la condición del estado asignado
            if (listaEstudiantes[i].getEstado().equals("Aprobado")) {
                aprobados++;
            } else {
                reprobados++;
            }
        }
        System.out.println("=======================================");

        // REQUERIMIENTO 5 y 6: Mostrar métricas de aprobación
        System.out.println("\n=== ESTADISTICAS DEL GRUPO ===");
        System.out.println("Estudiantes Aprobados (Promedio >= 7.00): " + aprobados);
        System.out.println("Estudiantes Reprobados (Promedio < 7.00): " + reprobados);
        System.out.println("==============================");
        
        teclado.close(); // Se cierra el Scanner al terminar el programa
    }
}

File: Cpp/Ape4Ejercicio.cpp
#include <iostream>
#include <string>

using namespace std;

// ==========================================
// CLASE: Estudiante
// ==========================================
class Estudiante {
private:
    // Atributos encapsulados del estudiante
    string cedula;
    string nombre;
    string apellido;
    double nota1;
    double nota2;
    double nota3;
    double promedio;
    string estado;

public:
    // Constructor por defecto necesario para inicializar el arreglo en memoria
    Estudiante() {
        cedula = ""; nombre = ""; apellido = "";
        nota1 = 0.0; nota2 = 0.0; nota3 = 0.0;
        promedio = 0.0; estado = "Reprobado";
    }

    // Constructor parametrizado para asignar valores directamente
    Estudiante(string c, string n, string a, double n1, double n2, double n3) {
        cedula = c;
        nombre = n;
        apellido = a;
        nota1 = n1;
        nota2 = n2;
        nota3 = n3;
        calcularPromedio();
        determinarEstado();
    }

    // Métodos Getters y Setters para acceso seguro a los atributos
    string getCedula() { return cedula; }
    void setCedula(string c) { cedula = c; }

    string getNombre() { return nombre; }
    void setNombre(string n) { nombre = n; }

    string getApellido() { return apellido; }
    void setApellido(string a) { apellido = a; }

    double getNota1() { return nota1; }
    void setNota1(double n1) { nota1 = n1; calcularPromedio(); determinarEstado(); }

    double getNota2() { return nota2; }
    void setNota2(double n2) { nota2 = n2; calcularPromedio(); determinarEstado(); }

    double getNota3() { return nota3; }
    void setNota3(double n3) { nota3 = n3; calcularPromedio(); determinarEstado(); }

    double getPromedio() { return promedio; }
    string getEstado() { return estado; }

    // REQUERIMIENTO 3: Calcular automáticamente el promedio
    void calcularPromedio() {
        promedio = (nota1 + nota2 + nota3) / 3.0;
    }

    // NUEVA ESPECIFICACIÓN: El estudiante aprueba ÚNICAMENTE si el promedio es >= 7.00
    void determinarEstado() {
        if (promedio >= 7.00) {
            estado = "Aprobado";
        } else {
            estado = "Reprobado";
        }
    }

    // Imprime de forma ordenada los datos individuales del estudiante
    void mostrarInformacion() {
        cout << "Cedula: " << cedula << " | Alumno: " << apellido << " " << nombre 
             << " | Notas: [" << nota1 << ", " << nota2 << ", " << nota3 << "]"
             << " | Promedio: " << promedio << " | Estado: " << estado << "\n";
    }
};

// ==========================================
// FUNCIONES AUXILIARES DE VALIDACIÓN
// ==========================================

// REQUERIMIENTO 7: Validar que las notas estén estrictamente entre 0 y 10
double solicitarNotaValidada(string numeroNota) {
    double nota;
    while (true) {
        cout << "  Ingrese la " << numeroNota << " (0 - 10): ";
        if (cin >> nota && nota >= 0.0 && nota <= 10.0) {
            return nota; // Retorna la nota solo si es válida
        } else {
            cout << "  [Error] Nota invalida. Debe ser un numero entre 0 y 10. Intente de nuevo.\n";
            cin.clear(); // Limpia el flag de error de cin (evita bucles infinitos por letras)
            cin.ignore(10000, '\n'); // Descarta el buffer incorrecto
        }
    }
}

// ==========================================
// FUNCIÓN PRINCIPAL
// ==========================================
int main() {
    // REQUERIMIENTO 1: Registrar mínimo 5 estudiantes
    const int TOTAL_ESTUDIANTES = 5;
    Estudiante listaEstudiantes[TOTAL_ESTUDIANTES];

    int aprobados = 0;
    int reprobados = 0;

    cout << "=== SISTEMA DE REGISTRO DE ESTUDIANTES ===\n\n";

    // Bucle para el ingreso de datos de cada uno de los estudiantes
    for (int i = 0; i < TOTAL_ESTUDIANTES; i++) {
        string cedula, nombre, apellido;
        double n1, n2, n3;

        cout << "Estudiante #" << (i + 1) << ":\n";
        cout << "  Cedula: ";
        cin >> cedula;
        cout << "  Nombre: ";
        cin >> nombre;
        cout << "  Apellido: ";
        cin >> apellido;

        // REQUERIMIENTO 2: Ingresar las 3 notas usando la función de validación
        n1 = solicitarNotaValidada("nota 1");
        n2 = solicitarNotaValidada("nota 2");
        n3 = solicitarNotaValidada("nota 3");
        cout << "-------------------------------------------\n";

        // Asignación de datos al objeto correspondiente del arreglo
        listaEstudiantes[i] = Estudiante(cedula, nombre, apellido, n1, n2, n3);
    }

    // REQUERIMIENTO 4: Mostrar el listado completo de estudiantes registrados
    cout << "\n=== LISTADO COMPLETO DE ESTUDIANTES ===\n";
    for (int i = 0; i < TOTAL_ESTUDIANTES; i++) {
        listaEstudiantes[i].mostrarInformacion();

        // Conteo de estudiantes según la condición del estado asignado (>= 7.00)
        if (listaEstudiantes[i].getEstado() == "Aprobado") {
            aprobados++;
        } else {
            reprobados++;
        }
    }
    cout << "=======================================\n";

    // REQUERIMIENTO 5 y 6: Mostrar métricas cuantitativas de aprobación
    cout << "\n=== ESTADISTICAS DEL GRUPO ===" << endl;
    cout << "Estudiantes Aprobados (Promedio >= 7.00): " << aprobados << endl;
    cout << "Estudiantes Reprobados (Promedio < 7.00): " << reprobados << endl;
    cout << "==============================" << endl;

    return 0;
}